### ISSUE TITLE
Git version from semver not saved to package-lock.json

### ISSUE THREAD

**Issue by jharris4** [npm/npm#17245]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ x] npm is producing an incorrect install.
  - [ ]x npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
When I use a git semver for a dependency in package.json like this:
`"thepackage": "git+https://gitblit.mydomain.com/r/thepackage.git#semver:^1.2.3"`
and then
`npm install`

I get the following in the package-lock.json:
```
"thepackage": {
      "version": "git+https://gitblit.mydomain.com/r/thepackage.git#560c4e6ea4835228d38872ff8e615b075ef2d88c",
      "dev": true
    }
```

I can manually verify that the commitish stored in package-lock.json is correct, but that's a bit cumbersome. Would it not be possible to also store the version in package-lock.json to be able to verify it?

I'd expect to see something like the following:
```
"thepackage": {
      "version": "git+https://gitblit.mydomain.com/r/thepackage.git#1.2.3",
      "dev": true
    }
```

#### How can the CLI team reproduce the problem?

See above. Install a git package using semver. I'm using the gitblit open source git server, but the problem may be the same with github.

### supporting information:

 - `npm -v` prints: 5.0.3
 - `node -v` prints: 8.1.2
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: MacOS 10.12.5
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by frediana** [comment]
Agree, I have the same concern.

**Comment by zkat** [comment]
This is as designed: `package-lock.json` always locks down dependencies to _specific commits_. Rather than trust that people won't just change those refs, we resolve them fully.

So, it is incredibly unlikely for this to change. I think what we're doing now is definitely the right thing to do. You can quickly verify which (tag) version was resolved using `git describe <committish>`. That will give you the tag associated with it.

**Comment by jharris4** [comment]
Ok, I understand the reason for using the committish as the source of truth.

But losing the ability to check which actual semver version is referenced by the package-lock.json at a glance is a big drawback in my opinion. Especially given that the feature in npm 5 is advertised as "semver support for git dependencies".

If the committish is being resolved from a semver version tag then also ouputting that version to the lockfile has no downside that I can see. The committish would still be there...

Having to copy and paste the committish to a command line and run a command is nowhere close to convenient.

For example, we use commit diffs as a sanity check when upgrading dependencies, and committish are useless in that scenario.

I apologize if I'm sounding overly critical. It's really awesome that git semver support was added!

I'm curious, would you be open to evaluating a PR to add the version to package-lock.json if it served a purely informational purpose for git dependencies?

**Comment by zkat** [comment]
`resolved` is meant to be an *unambiguous* and *immutable* pointer to the data we want to fetch. Or at least, as close to that as possible. With git, neither tags nor branches are immutable. Not to mention we'd otherwise have no way to check integrity itself.

I think I'd consider something that adds metadata to [`pacote`](https://npm.im/pacote), for sure. Something like a `_resolvedGitTag`, which is only present if we did semver-range resolution. Once that's in place, I think it might be doable to add an informational field to `package-lock.json`. I'm gonna /cc @iarna to get a sense of how she feels about this change, too.

I'll reopen this issue for now.

**Comment by jharris4** [comment]
I agree that `resolved` needs to be unambiguous and immutable. But part of my confusion here is the fact that `resolved` *isn't even used* for git dependencies. So perhaps I should have written in my initial comment in this issue that I would expect the `package-lock.json` entry for a git dependency to look something like this:
```
"thepackage": {
  "version": "1.2.3",
  "resolved": "git+https://gitblit.mydomain.com/r/thepackage.git#560c4e6ea4835228d38872ff8e615b075ef2d88c",
  "dev": true
}
```

**Comment by zkat** [comment]
@jharris4 we probably won't be moving the full git url out of `version`, fwiw. That is a breaking change to the package-lock spec, and this does not seem to be worth that cost. But it's ok to bikeshed about what the new, ignorable field will be.

**Comment by jharris4** [comment]
yeah, too late for that kind of change, although I am curious as to how the git url + committish ended up in the version field to begin with... :-)

My vote would've been for using `version` to store the version number, but since that's taken, I'd say whatever makes the most sense to you guys given the internals.

fwiw I would be happy to look at making a PR if it'd help, although a pointer in the right direction would be super handy.

**Comment by zkat** [comment]
@jharris4 https://github.com/zkat/pacote/blob/latest/lib/fetchers/git.js#L132-L147
